test/membrane_opus/encoder_test.exs

```diff
@@ -0,0 +1,60 @@
+defmodule Membrane.Opus.Encoder.EncoderTest do
+  use ExUnit.Case, async: true
+
+  alias Membrane.Opus.{Encoder, Decoder}
+  alias Membrane.Caps.Audio.Raw
+  import Membrane.Testing.Assertions
+  alias Membrane.Testing
+  alias Membrane.Element.File
+  import Membrane.ParentSpec
+
+  @input_path "test/fixtures/raw_packets"
+
+  setup do
+    elements = [
+      source: %File.Source{
+        location: @input_path
+      },
+      encoder: %Encoder{
+        application: :audio,
+        input_caps: %Raw{
+          channels: 2,
+          format: :s16le,
+          sample_rate: 48_000
+        }
+      },
+      decoder: %Decoder{
+        sample_rate: 48_000,
+        channels: 2
+      },
+      sink: Testing.Sink
+    ]
+
+    links = [
+      link(:source)
+      |> to(:encoder)
+      |> to(:decoder)
+      |> to(:sink)
+    ]
+
+    {:ok, pipeline_pid} =
+      Testing.Pipeline.start_link(%Testing.Pipeline.Options{
+        elements: elements,
+        links: links
+      })
+
+    {:ok, %{pipeline_pid: pipeline_pid}}
+  end
+
+  # NOTE: a full integration test of Opus will require implementation of an
+  # Opus container and parser. The libopus encoder tests do the same thing
+  # we're doing here: just ensuring that encoded packets can be decoded.
+  #
+  # https://gitlab.xiph.org/xiph/opus/-/blob/master/tests/test_opus_encode.c#L160
+  test "eating our own dog food", context do
+    %{pipeline_pid: pipeline_pid} = context
+    Membrane.Pipeline.play(pipeline_pid)
+    assert_start_of_stream(pipeline_pid, :sink)
+    assert_end_of_stream(pipeline_pid, :sink)
+  end
```

Comment:
I think that we have to check somehow that after encoding and decoding data remains unchanged. At this moment we only ensure that encoding and decoding process has been finishd but we don't know if it is proper.

Reply:
Unfortunately that won't work. Since Opus is a lossy codec, the output will not be the same. We also can't test that a raw stream, when converted to Opus, will be the same as the reference Opus file because the codec is also VBR by default, so it's not deterministic. This is why I believe libopus does their testing exactly like this.

It _might_ be possible to try and use a fixed bit rate and see whether the output is deterministic? LMK what you think :)

Reply:
Actually, the data may change after encoding and decoding, as Opus is a lossy codec. We need to check if the data is encoded properly though, for example using a reference file.

Reply:
Yeah, my bad. We obviously cannot encode decode and compare.

Reply:
@ledhed2222 didn't know Opus is non-deterministic, though we can still compare sizes (in bytes) of input and output audio.

Reply:
that's a good idea @mat-hek. let me confirm with someone at opus that it is indeed non-deterministic in VBR mode. that was what I assumed looking at their tests, which only validate decoded streams against a reference

Reply:
ok you're right - libopus itself DOES produce deterministic encoding. i'll fix up these tests

Reply:
i'm really struggling to get a reference opus stream. i can't just use the one that's already in the repo without knowing what its encoding settings were. help!

Reply:
`ffmpeg -f lavfi -i sine=f=440 -acodec libopus sample.opus`

Reply:
left a message in discord too. i could use some help here. my encoded flat files are significantly larger than the references. i've looked up and down my implementation for places where i may have errors but i'm not seeing anything.

Reply:
How do they look like after decoding, compared to the references?